### ISSUE TITLE
Proposal: Design Tokens (WIP)

### ISSUE THREAD

**Issue by notlee** [Financial-Times/origami#242]
# Introduce Design Tokens

**Draft: This document is more of a brain dump at the moment. There is a style dictionary experiment at [Financial-Times/origami-design-tokens-style-dictionary-experiment](https://github.com/Financial-Times/origami-design-tokens-style-dictionary-experiment/)**

Design tokens are a method to store, document, and distribute design decisions across platforms; for example brand colour, spacing, and typographic scale values across web, native apps, and design tooling. There is a growing ecosystem of methodologies and tools around design tokens to support team collaboration and new product development.

## Motivation

**Draft: These are possibilities, not refined motivations.**

- To provide a mechanism to introduce dark mode branding to Origami. The app currently implements dark mode but it’s difficult to implement at the app level in a consistent way as styles must be overridden, which would be [time consuming and unreliable](https://origami.ft.com/docs/components/customisation/#avoid-css-overrides). The app instead relies on inverting everything and then opting out some elements (images, slate backgrounds). This is limited. As Mark Limb, designer on the apps team, writes "the brand is completely lost and in some places the colours become meaningless, often distracting, [...] I believe we do need a dark variant of our FT palette".
- To make it easier to customise Origami by adding new brands or variants with brands.
    1. FT Creative use Origami components but heavily modify them, this is difficult, unsupported, and can be unreliable. There's an opportunity to work with FT Creative to meet their requirements in a global design system and add suitable component variants to Origami.
    2. Specialist Titles: Used to use Origami components more but now largely have their own design system, with some exceptions like the [o-cookie-message](https://registry.origami.ft.com/components/o-cookie-message@6.0.0) Origami component which is still in use. There's an opportunity for us to collaborate more (share code and design thinking) to reduce repeated work and improve design consistency. Where Origami components are used across specialist titles, ft.com, and internal tools (like [o-cookie-message](https://registry.origami.ft.com/components/o-cookie-message@6.0.0)), we should look to design component variants in Origami specifically for Specialist Titles to reinforce the title’s brand.
- To share platform agnostic design decisions. Currently many design decisions are encoded as Sass variables in Origami components (e.g. the spacing between 2 elements, the colour of "paper", the colour of a link, our typographic scale, size of headings). These values need to be duplicated across platforms, which increases technical complexity and introduces opportunities for error:
    1. In a component's JavaScript, or passed along in [some other non-standard means](https://css-tricks.com/making-sass-talk-to-javascript-with-json/) (since we cannot rely on CSS Custom Properties and maintain IE11 support).
    2. In native app elements. As well as being copied over they may need to be converted to a different format, too.
    3. In design tooling (e.g. Figma).
- To improve designer and engineer communication, by using the same terminology/values across developer and designer tooling more often.
- Make it easier to find and edit values which represent design decisions. Currently values are spread across multiple component repositories, with multiple formats, across platform specific Sass and JavaScript.
- Simplify component Sass by abstracting complex logic to a design token compile step. For example the complex logic used to [calculate the colour of a button](https://github.com/Financial-Times/o-buttons/blob/v7.0.0/src/scss/_functions.scss#L144), this is hard to reason about or improve in Sass.
- To distribute tint-able FT icons to native apps?

## explanation

> Provide a detailed explanation of your proposal
> This section is required

## work required

> Add here steps required to make your proposal a reality
> Provide links to the code that would need to change if relevant
> This section isn't required to open a proposal, but is required before a proposal can be accepted

## alternatives

> Provide some alternatives to your proposal and their benefits and drawbacks
> This section is required

## research notes

Reading list 👀

* [GitHub - sturobson/Awesome-Design-Tokens: A list of resources on all things to do with Design Tokens](https://github.com/sturobson/Awesome-Design-Tokens)
* [CSS Tricks: What are design tokens](https://css-tricks.com/what-are-design-tokens/)
* [Design Token Community Group (glossary)](https://www.designtokens.org/glossary/)
* [Bradfrost Design token architecture](https://bradfrost.com/blog/link/building-better-products-with-a-design-token-pipeline/)
* https://mobile.twitter.com/svorklab/status/141138043182460109
* [Design token 101 talk](https://youtu.be/G4mui3OCMVA?t=5376)
* [Dark Mode with Style Dictionary | dbanksdesign](https://dbanks.design/blog/dark-mode-with-style-dictionary/)
* [Announcement: Adobe XD extension for Visual Studio Code and Design System Packages (DSP)](https://vimeo.com/466431987)(invites guests from other tools/platforms to talk)
* [Specify: Introduction to design tokens](https://specifyapp.com/blog/introduction-to-design-tokens)

Related tools / platforms 👀

- [Theo](https://github.com/salesforce-ux/theo)
- [Style Dictionary](https://amzn.github.io/style-dictionary/#/)
- [Diez](https://diez.org/)
- [Specify](https://specifyapp.com/)
- [Adobe Design System Package](https://github.com/AdobeXD/design-system-package-dsp) (DSP)
- [Knapsack](https://www.knapsack.cloud/)
- [InVision Design System Manager](https://www.invisionapp.com/design-system-manager)

## notes

> Add notes about unresolved questions or other things that don't fit well into the sections above
> This section is optional

**Comment by notlee** [comment]
Draft proposal moved from PR: https://github.com/Financial-Times/origami/pull/129

**Comment by notlee** [comment]
Ooops, `@todo` just get a few notifications

**Comment by notlee** [comment]
Please review and comment team. We have a Q1 OKR in place to replace o-brand with design tokens internally, for 50% of components; and are talking with the apps team about spiking dark mode – looking to move that to a future Q.

**Comment by JakeChampion** [comment]
I agree with the proposed approach as it is small in it's scope and allows us to get real-world experience in using design tokens within figma and origami code 👍

**Comment by chee** [comment]
i found this sentence confusing to read:
>  Plus, Sass variables may be autocompleted by text editors where the arguments for Origami's existing Sass functions cannot be reliably.

what does cagoule mean here?
> Publish IOS asset cagoule

There's a lot of context in this proposal, and multiple lists of things to do, and I'm not clear on what work accepting this proposal actually commits us to?

**Comment by notlee** [comment]
Thanks for taking a look @chee!

>i found this sentence confusing to read

Did you figure out the meaning? I'm open to suggestions for phrasing that line differently. 

>what does cagoule mean here?

Oh yes, I forgot to define that. I've added a link to the [Apple docs](https://help.apple.com/xcode/mac/current/#/dev10510b1f7). 

This is a helpful summary: "While Android has resources, iOS has assets. iOS has image and color assets, which can support color modes (and have more modes than just light and dark too). Assets are stored in an asset catalog which is a folder with sub-folders that have a specific extension (.colorset or .imageset) and a JSON file named Contents.json." - https://dbanks.design/blog/dark-mode-with-style-dictionary/

>There's a lot of context in this proposal, and multiple lists of things to do, and I'm not clear on what work accepting this proposal actually commits us to?

The "Proposed Approach" heading is the course of actions this proposal commits us to.

**Comment by chee** [comment]
> > what does cagoule mean here?

> Oh yes, I forgot to define that. I've added a link to the Apple docs.

I think maybe there is a typo there. The proposal says "cagoule" which i think is a kind of hooded anorak and sometimes when they are rolled up the hood (or front pocket) doubles as a bag into which the cagoule can be packed but the website says "catalog"

> The "Proposed Approach" heading is the course of action this proposal commits us to.

okay this looks good to me

**Comment by notlee** [comment]
The proposed approach has been dropped. We’re working on getting buy-in on initiatives such as improving Origami’s multi-brand / theme / platform support so we can grow the team to support continued work. The investigation and [design token exploration](https://github.com/Financial-Times/origami/pull/360) will be helpful to feed into that
